src/v2/providers/alerts/billing.ts

```diff
@@ -0,0 +1,106 @@
+
+/** Data */
+/** Payload for plan update */
+export interface PlanUpdatePayload {
+  ['@type']: 'com.google.firebase.firebasealerts.PlanUpdatePayload';
+  billingPlan: string;
+  principalEmail: string;
+}
+/** Payload for plan automated update */
+export interface PlanAutomatedUpdatePayload {
+  ['@type']: 'com.google.firebase.firebasealerts.PlanAutomatedUpdatePayload';
+  billingPlan: string;
+}
+
+/** @internal */
+export const planUpdateAlert = 'billing.planUpdate';
+/** @internal */
+export const automatedPlanUpdateAlert = 'billing.automatedPlanUpdate';
+
+/** Cloud Event Type */
+interface WithAlertType {
+  alertType: string;
+}
+export type BillingEvent<T> = CloudEvent<FirebaseAlertData<T>, WithAlertType>;
+
+/** Handlers */
+/** Handle a plan update published */
+export function onPlanUpdatePublished(
+  handler: (event: BillingEvent<PlanUpdatePayload>) => any | Promise<any>
+): CloudFunction<FirebaseAlertData<PlanUpdatePayload>>;
+export function onPlanUpdatePublished(
+  opts: options.EventHandlerOptions,
+  handler: (event: BillingEvent<PlanUpdatePayload>) => any | Promise<any>
+): CloudFunction<FirebaseAlertData<PlanUpdatePayload>>;
+export function onPlanUpdatePublished(
+  optsOrHandler:
+    | options.EventHandlerOptions
+    | ((event: BillingEvent<PlanUpdatePayload>) => any | Promise<any>),
+  handler?: (event: BillingEvent<PlanUpdatePayload>) => any | Promise<any>
+): CloudFunction<FirebaseAlertData<PlanUpdatePayload>> {
+  return onOperation<PlanUpdatePayload>(
+    planUpdateAlert,
+    optsOrHandler,
+    handler
+  );
+}
+
+/** Handle an automated plan update published */
+export function onAutomatedPlanUpdatePublished(
+  handler: (
+    event: BillingEvent<PlanAutomatedUpdatePayload>
+  ) => any | Promise<any>
+): CloudFunction<FirebaseAlertData<PlanAutomatedUpdatePayload>>;
+export function onAutomatedPlanUpdatePublished(
+  opts: options.EventHandlerOptions,
+  handler: (
+    event: BillingEvent<PlanAutomatedUpdatePayload>
+  ) => any | Promise<any>
+): CloudFunction<FirebaseAlertData<PlanAutomatedUpdatePayload>>;
+export function onAutomatedPlanUpdatePublished(
+  optsOrHandler:
+    | options.EventHandlerOptions
+    | ((event: BillingEvent<PlanAutomatedUpdatePayload>) => any | Promise<any>),
+  handler?: (
+    event: BillingEvent<PlanAutomatedUpdatePayload>
+  ) => any | Promise<any>
+): CloudFunction<FirebaseAlertData<PlanAutomatedUpdatePayload>> {
+  return onOperation<PlanAutomatedUpdatePayload>(
+    automatedPlanUpdateAlert,
+    optsOrHandler,
+    handler
+  );
+}
+
+/** @internal */
+export function onOperation<T>(
+  alertType: string,
+  optsOrHandler:
+    | options.EventHandlerOptions
+    | ((event: BillingEvent<T>) => any | Promise<any>),
+  handler: (event: BillingEvent<T>) => any | Promise<any>
+): CloudFunction<FirebaseAlertData<T>> {
+  if (typeof optsOrHandler === 'function') {
+    handler = optsOrHandler as (event: BillingEvent<T>) => any | Promise<any>;
```

Comment:
Woo `(event: BillingEvent<T>) => any | Promise<any>;` is used thrice! This meets the threshold for DRY for me:

```ts
type BillingEventHandler<T> = (event: BillingEvent<T>) => any | Promise<any>;
```

Reply:
Oh that looks way better, thanks for the suggestion!

Reply:
@taeold Removed based on typing feedback below